### PR TITLE
Pdf uri extractor and pivoting

### PR DESCRIPTION
**Pull request by federicofantini**
# Description
Added pdf uri extractor (analyzer) and dowload file from uri (analyzer) + pivot and playbook configs (migrations).

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [X] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [X] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [X] The pull request is for the branch `develop`
- [X] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [X] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [ ] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [X] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [X] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [ ] If you created a new analyzer and it is free (does not require any API key), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [ ] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [ ] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
    - [ ] If the plugin interacts with an external service, I have created an attribute called precisely `url` that contains this information. This is required for Health Checks. 
    - [ ] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [ ] I have added that raw JSON sample to the `MockUpResponse` of the `_monkeypatch()` method. This serves us to provide a valid sample for testing.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [ ] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.
- [ ] After you had submitted the PR, if `DeepSource`, `Django Doctors` or other third-party linters have triggered any alerts during the CI checks, I have solved those alerts.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

### CODE REVIEW

**Comment by federicofantini** [comment]
image

**Comment by federicofantini** [comment]
image

File: api_app/analyzers_manager/observable_analyzers/download_file_from_uri.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        else:
+            if filename := custom_slugify(
+                os.path.basename(urlparse(unquote(self.observable_name)).path)
+            ):
+                with open(os.path.join(self.basefolder, filename), "wb") as tmp:
+                    tmp.write(r.content)
+                    tmp.flush()
+                    tmp.close()
+                    result["stored_in"] = os.path.join(self.basefolder, filename)
+        return result
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/download_file_from_uri.py:77]: I am afraid that writing the file to disk would leave us with a lot of possible complication regarding security and permissions.
In my opinion the best option would be to save the file blob as base64 in a specific field and then decode it from the pivot at the time when you have to create a new file analysis

**Review by mlodic — CHANGES_REQUESTED** [review]

File: api_app/analyzers_manager/file_analyzers/pdf_info.py [hunk 2]
```diff
@@ -19,14 +22,64 @@ def flatten(list_of_lists: List[List[Any]]) -> List[Any]:
...
+        if not pdfreader_success and not (pdfid_success and peepdf_success):
+            raise AnalyzerRunException("all readers failed")
+
+        # pivot uris in the pdf only if we have one page
+        if self.results["pdfreader"]["pages"] == 1:
+            uris = []
+            for s in self.results["peepdf"]["stats"]:
+                uris.extend(s["uris"])
+            uris.extend(self.results["pdfreader"]["uris"])
+            logger.info(f"{uris=}")
```
Comment (mlodic) [inline api_app/analyzers_manager/file_analyzers/pdf_info.py:41]: add context info in the log like the pdf md5

File: api_app/analyzers_manager/file_analyzers/pdf_info.py [hunk 3]
```diff
@@ -19,14 +22,64 @@ def flatten(list_of_lists: List[List[Any]]) -> List[Any]:
...
+            try:
+                doc = pdfreader.PDFDocument(fd)
+            except pdfreader.exceptions.ParserException:
+                parser_exception = True
+            else:
+                self.results["pdfreader"]["pages"] = doc.root["Pages"]["Count"]
+                for page in doc.root["Pages"]["Kids"]:
+                    for annot in page["Annots"]:
+                        if "A" in annot:
+                            if "URI" in annot["A"]:
```
Comment (mlodic) [inline api_app/analyzers_manager/file_analyzers/pdf_info.py:71]: merge

**Pull request merged**